File: presenterRx/src/main/java/me/ameriod/lib/mvp/presenter/rx/IObservableSchedulerRx.java
package me.ameriod.lib.mvp.presenter.rx;

import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action0;
import rx.schedulers.Schedulers;

public interface IObservableSchedulerRx {

    <T> Observable.Transformer<T, T> scheduleObservable();

    IObservableSchedulerRx SUBSCRIBE_COMPUTATION_OBSERVE_ANDROID_MAIN = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.computation())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            };
        }
    };

    IObservableSchedulerRx SUBSCRIBE_IO_OBSERVE_ANDROID_MAIN = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread());
                }
            };
        }
    };

    IObservableSchedulerRx IO = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.io())
                            .observeOn(Schedulers.io());
                }
            };
        }
    };

    IObservableSchedulerRx COMPUTATION = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.computation())
                            .observeOn(Schedulers.computation());
                }
            };
        }
    };

    IObservableSchedulerRx IMMEDIATE = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.immediate())
                            .observeOn(Schedulers.immediate());
                }
            };
        }
    };

    IObservableSchedulerRx TRAMPOLINE = new IObservableSchedulerRx() {
        @Override
        public <T> Observable.Transformer<T, T> scheduleObservable() {
            return new Observable.Transformer<T, T>() {
                @Override
                public Observable<T> call(Observable<T> observable) {
                    return observable
                            .subscribeOn(Schedulers.trampoline())
                            .observeOn(Schedulers.trampoline());
                }
            };
        }
    };
}

File: presenterRx/src/main/java/me/ameriod/lib/mvp/presenter/rx/IObservableSchedulerRx.kt
package me.ameriod.lib.mvp.presenter.rx

import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.functions.Action0
import rx.schedulers.Schedulers

interface IObservableSchedulerRx {

    fun <T> scheduleObservable(): Observable.Transformer<T, T>

    companion object {

        val SUBSCRIBE_COMPUTATION_OBSERVE_ANDROID_MAIN: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.computation())
                            .observeOn(AndroidSchedulers.mainThread())
                }
            }
        }

        val SUBSCRIBE_IO_OBSERVE_ANDROID_MAIN: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                }
            }
        }

        val IO: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.io())
                            .observeOn(Schedulers.io())
                }
            }
        }

        val COMPUTATION: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.computation())
                            .observeOn(Schedulers.computation())
                }
            }
        }

        val IMMEDIATE: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.immediate())
                            .observeOn(Schedulers.immediate())
                }
            }
        }

        val TRAMPOLINE: IObservableSchedulerRx = object : IObservableSchedulerRx {
            override fun <T> scheduleObservable(): Observable.Transformer<T, T> {
                return Observable.Transformer<T, T> { observable ->
                    observable
                            .subscribeOn(Schedulers.trampoline())
                            .observeOn(Schedulers.trampoline())
                }
            }
        }
    }
}
